feat: job configuration subsystem

sbliven opened the pull request:
## Description
This culminates the jobs release, a major change to how Jobs work in scicat. This allows flexible interaction with other facility infrastructure, including archival and publishing services. Rather than a few hard-coded job types, jobs can now be configured using the `JOB_CONFIGURATION_FILE` env variable, which points to a yaml file allowing customization how scicat handles new and updated jobs.

Full documentation is provide in https://github.com/SciCatProject/documentation/pull/56. Please read that for information about authorization, configuration, and options.

The following actions are included in the initial release, with additional actions planned to follow:
- url
- rabbitmq
- validate
- email
- log

## Fixes

- #1668
- #1439
- #1438
- #1435
- #1424
- #1411
- #1388
- #1341
- #1294
- #1265
- #1154
- #1152
- #1124
- #1123
- #1121
- #1119
- #1114
- #1113
- #841
- #840
- #839
- #838
- #740
- #626
- #625
- #624
- #623
- #622
- #621
- #620
- #1754 


## Tests included

- [x] Included for each change/fix?
- [x] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

### official documentation info
https://github.com/SciCatProject/documentation/pull/56

nitrosx committed:
wip of the day implementing jobs

Junjiequan committed:
fix: updated mathJS package version && improved number check

nitrosx committed:
work in progress from scicat marathon in CPH 2023/10/24-25

HayenNico committed:
Initial suggestions #622

CreateJobDTO and UpdateJobDTO with API decorators etc.. Besides basic functionality already required in the controller some additional suggestions for properties.

HayenNico committed:
fix: removed unnecessary imports

HayenNico committed:
Merge branch 'master' into Rel-4-4-Issue-622

HayenNico committed:
fix eslint complaints

HayenNico committed:
fix eslint complaints

HayenNico committed:
fix typo in job id decription

HayenNico committed:
fix typo in job id decription

HayenNico committed:
Merge branch 'Rel-4-4-Issue-622' of https://github.com/SciCatProject/scicat-backend-next into Rel-4-4-Issue-622

HayenNico committed:
replace update-job.dto with update-jobstatus.dto

Removed update-job.dto as it will not be implemented.
Added job-statusupdate.dto.

HayenNico committed:
Merge pull request #844 from SciCatProject/Rel-4-4-Issue-622

Release - Jobs DTO (#622)

Commit:
reviewed version of branch issue621_rel4_4 (by myself, NEEDS REVIEW)

nitrosx committed:
Update job.schema.ts

commented schema example

Commit:
remove optional id from datasets

nitrosx committed:
job schema discussion and definition

nitrosx committed:
resolved conflicts

Commit:
finish off, leave things out for later

nitrosx committed:
Merge branch 'release-4-4' into release_jobs_issue621_schema

nitrosx committed:
Merge pull request #905 from SciCatProject/release_jobs_issue621_schema

Release jobs, defining job schema

nitrosx committed:
merged latest changes from release-4-4, fixed import in job schema, added job auth enumeration

HayenNico committed:
Initial refactor for job creation and status updates

Updated service functions for job creation, status updates and querying.

HayenNico committed:
Change sendJobStartEmail

Suggestion for the change to the email at job start.
Old version was very specific to archive/retrieve workflows, which should now go to separate microservices instead.

HayenNico committed:
Minor refactor job.schema

nitrosx committed:
updated jobs module

nitrosx committed:
Merge pull request #923 from SciCatProject/issue-623

updated jobs module

nitrosx committed:
work from the meeting on 2023/12/06

nitrosx committed:
Merge pull request #938 from SciCatProject/release-jobs-meeting-20231206

work from the meeting on 2023/12/06

HayenNico committed:
Remove custom email code from jobs.service

HayenNico committed:
Fix return types from jobs.service

Return types from jobs.service actions now match equivalent operations for datasets.service

HayenNico committed:
Merge pull request #907 from SciCatProject/release-jobs-service-issue624

Release Jobs - Jobs Service Refactor (#624)

HayenNico committed:
Merge branch 'master' into release-jobs

sbliven committed:
Initial job config implementation (squashed)

- config/jobconfig.ts provides data types and code for reading the
  configuration json
- config/actions implement basic log and email actions
- Fix some compilation errors (more remain)
- Includes a unit test for config parsing
- Implement job actions in the jobs.controller.ts

8 compilation errors remain. In particular, it's currently inconsistent whether `ActionJob.actionType` should be static.

nitrosx committed:
Merge pull request #1002 from sbliven/issue-841-squashed

Job Config (squashed)

sbliven committed:
Fix the static actionType compilation errors

- JobActionClass specifies static properties required to register the
  action (eg `static actionType` property)
- JobAction specifies instance properties. It adds `getActionType()`
  with the contract that it should return the same string.
- Move oneOrMore to util.ts

HayenNico committed:
Merge branch 'master' into release-jobs

sbliven committed:
Add 'JobOperation' layer in JobConfig data model

JobOperation represents create, read, updated, etc.

It primarily consists of a list of actions, but adds the ability to
configure authentication at this level rather than as a JobAction.

sbliven committed:
Run lint:fix

sbliven committed:
Revert job-example change

(but other things in this file still need to be updated)

despadam committed:
WIP create job

sbliven committed:
Minor changes documenting job authorization

Add jobConfig.json.example, which is intended to be customized and
renamed to jobConfig.json as the main job configuration file.

sbliven committed:
Rename Action enum to AuthOp

nitrosx committed:
Merge pull request #1021 from sbliven/jobconfig

Improve JobConfig

despadam committed:
schema validation

despadam committed:
fixes

despadam committed:
fixes + email action param validation

sbliven committed:
Load job configuration lazily as a singleton

despadam committed:
Update comments in authop.enum.ts: change individual to data instance authorization

sbliven committed:
Merge pull request #1080 from sbliven/839-create-job

839 create job

despadam committed:
Job creation: fix missing _id

despadam committed:
Merge pull request #1091 from SciCatProject/fix-job-create

Job creation: fix missing _id

nitrosx committed:
Update published-data.service.ts

backported change from master

sbliven committed:
Initial urlaction implementation

The test configuration simply calls /health and logs the result.

despadam committed:
Job statusUpdate basic implementation

sbliven committed:
Validate 'type' instead of 'schema' for job dto

despadam committed:
Merge pull request #1110 from SciCatProject/838-jobs-update

Job statusUpdate

nitrosx committed:
Merged master in release jobs

sbliven committed:
Add 'jsonify' Handlebars helper

sbliven committed:
Accept template for URLAction config

This is a restrictive implementation; for more on security, see
https://github.com/SciCatProject/scicat-backend-next/issues/1113

sbliven committed:
Merge  branch 'release-jobs' into urlaction

sbliven committed:
Merge pull request #1093 from SciCatProject/urlaction

Add URLAction. Trigger REST calls following job operations.

nitrosx committed:
merged master

despadam committed:
Jobs: statusHistory, instance auth based on config, Get functions (#1133)

* fixes based on PR 1110 comments

* store configuration version

* complete instance auth

* Adding '@Req() request: Request'

* restore jobParams in schema

* fix: check authenticated user & job creation message

nitrosx committed:
next draft of job authorization

nitrosx committed:
new version of status update dto

nitrosx committed:
merged current version of release-jobs in this branch

nitrosx committed:
applied improvements from comments

nitrosx committed:
fixed linting

Junjiequan committed:
Improved types & implemented job-create interceptor

Junjiequan committed:
refactor: type improvement for jobs-aurhotization PR

nitrosx committed:
Merge pull request #1202 from SciCatProject/jobs-authorization-refactor

refactor: type improvement for jobs-aurhotization PR

despadam committed:
remove statusHistory from job schema

nitrosx committed:
Merge pull request #1173 from SciCatProject/jobs-authorization

Jobs authorization

despadam committed:
Remove configVersion and rename statusUpdate variables (#1273)

* remove configVersion from job schema, fix statusUpdate vars, update GET/jobs

* remove additional filter from findAll, rename to StatusUpdateJobAuth and statusUpdateJobGroups

* move configVersion to the top of the config file

despadam committed:
Remove unused fields from job schema (#1307)

despadam committed:
Jobs should only use the SciCat instance configuration (#1320)

* use only global instance configuration

* remove print

despadam committed:
Job Status Update: Implement RabbitMQ Action (#1284)

* add RabbitMQJobAction, fix how job config is stored

* rabbitmq queue fixes

* fix comment

* fixes based on PR comments, register action for job create, send json

* finalize rammitmq

sofyalaski committed:
Jobs tests (#1286)

* first tests

* feat: reset dataset before every api test (#1131)

* reset dataset before every api test

* Clear collections before each api test

* Fixes after jobs testing:
1. Use CreateJobAuth enum instead of AuthOp to define jobDatasetAuthorization
2. Shift a block if groupOwner is present down after the admin check was passed (admin create a job for anonym)
3. Check if current authorization is in jobDatasetAuthorization in instanceAuthorizationJobCreate (types  mismatching)
4. Add a check if unauthorized user creates a job for someone else

* Fixes after jobs testing:
if anonymous user creates a job, initialize a jwt instance for him/her

* Fixes after testing:
string array includes string enum

* dataset(s)Validation redundancy after a merged PR

* test Jobs authorization wip

* fix filtering syntax

* fixes after Job testing:
1. comment out jobConfig matchin, as now in interceptor
2. fix filtering object hierarchy
in #dataset kind of jobConfig's
3. not defining jobInstance ownerUser/Group as empty string
4. add jobInstance.jobParams initialization
5. add a check if datasets passed in the ID list exist

* minor changes

* fix findOne filtering syntax

* add Jobs tests

* remove redundancy

* fix numbering

* extend UPDATE_JOB_GROUP user's right to match docs

* remove unneccessary arguments

* make function return value of findOne  an instance of JobClass

* add statusUpdate tests

* add jobParams to the job instance

* Patch tests

* Fixes after tests:
* Anonymous user can patch status update to a job in #all config. Added instance authorization

* Implement JobDelete authorization

* add deleteJobGroups from env to configuration

* add username for anonymous user when creating/updating status of the job

* statusUpdate instead of update

* Fixes after testing:

* similarly to Datasets, mongoose returns an Object and not JobClass instances. To pass this into CASL we first need  to make those class instances. generateJobInstanceForPermissions implements a copy of object as correct class instance only with properties required in casl.

* functionality of instanceAuthorizationJobStatusUpdate is now implemented in the patch endpoint function. This is done to avoid redefining a found job as a instance of JobClass with all properties

* For both GET endpoints only the endpoint authorisation was implemented for user in the functions of the controller. To add the instance authorization too, first create JobClass instance of the found job Objects and pass them to respective casl expressions.

* adds implementation of delete endpoints. The rules are defined in the CheckPolicies decorator.

* Finalized tests for authorization

* add delete_job_groups definition

* fix unit tests after test jobconfig.json changes

* empty collections before testing

* changes requested in PR

* eremove version duplicate

* change getJobMatchingConfiguration function to explicitly use job type, st it can be used for all cases where we need to extract the configuration, without having to pass the full dto

* run full testing workflow on pushing to release branches

* add env variable for jobs config path for github workflow

* Fix lint errors

- Run lint:fix
- Ignore no-explicit-any rule for jobParams (which are not type checked)
- remove 'read' jobOperation (not implemented)

* Fix lint following merge

* Update test/config/pretest.js to use "dotenv"

Co-authored-by: Jay <a331998513@gmail.com>

* Fix pretest.js after merging suggestions

* Fix pretest.js

* Fix jobconfig no-explicit-any issues

- Revert previous changes to ignore no-explicit-any
- Use `unknown` for jobParam inputs plus stronger runtime checks
- Fix compilation error in elastic-search service relating to
  deleting non-optional variables. 
- Remove rabbitmqaction validation. RabbitMQ should ignore the DTO body,
  and validate the config itself in the constructor.

* Fix jobconfig no-explicit-any issue (missed logaction)

* changes in tests:
0060 now checks if a job was passed without a required parameter JobParams
0065 (new) checks if a job was passed with an empty object for jobParams
1950 now tries to delete a job that doesn't exist and fails
1960 was called 1950 before
counts on GET are changed (because 0060 before was passing)

* address comments in the PR:
*  now makes one request to the db to find datasets and extracts their pids.
*  loggs the names of all dataset pids that are not in the db in case of an error
*  job endpoint implements the check that the jobParams were passed and are not empty, otherwise it will throw a bad request with respective message
*  job endpoint loggs the id of job to be deleted and implements a check if job exist in db, otherwise throws an error.

---------

Co-authored-by: Jay Quan <a331998513@gmail.com>
Co-authored-by: Spencer Bliven <spencer.bliven@gmail.com>
Co-authored-by: Despina Adamopoulou <16343312+despadam@users.noreply.github.com>

sbliven committed:
Improve URLAction templating (#1336)

* Add templating to URLAction

The jobs configured with a `url` action can now take templates for the
`url`, `body` and `headers` fields in the job configuration.

* Handlebar helpers: `urlencode` `base64enc` and `job_v3`

- Use `{{{urlencode var}}}` for URL components
- `base64enc` can be useful for passing data to services
- `job_v3` converts the current v4 job schema to a backwards-compatible
  form. For instance, use `body: "{{{jsonify (job_v3 this)}}}"` to pass
  a more backwards-compatible json body.

* fix linting

---------

Co-authored-by: Despina Adamopoulou <despoina.adamopoulou@empa.ch>

despadam committed:
Validate actions during their creation (#1354)

* remove validation from actions, store only configVersion inside each job

* remove action validation from testing

* fix testing

* restore configuration in the schema, fix tests

* fix linting

* fix error message

despadam committed:
Store configVersion in the job schema (#1364)

* update job schema

* store configVersion during job creation

* fixes in controller

* rename variable

* remove job configuration in jobInstance to enable casl authorization (#1395)

* wip cherry-picking commits

* wip: casl for job instances before adding argument with jobConfig

* merged cherry-picked commit that splits casl, jobs authorization requires two arguments, the new being the jobConfig property. Casl for jobs instances is restructred to use it in the if statements instead of configuration being part of jobInstance

* fix linting

* proposals pass tests

---------

Co-authored-by: Sofya Laskina <44316816+sofyalaski@users.noreply.github.com>

despadam committed:
merge master to release-jobs

despadam committed:
Fix error after merging

despadam committed:
fix beforeEach in tests after merging

despadam committed:
Add jobResultObject field (#1410)

* add jobResultObject field

* fix linting

* fix optional values in statusUpdate dto

despadam committed:
Jobs: templating for email action (#1326)

* email action wip

* fix email templating

* remove unnecessary file

* add optional auth field

* rename field to bodyTemplateFile

* fix lint errors

sbliven committed:
Add JobAction.validate method.

This partially reverts commit b218952495b36d836c155e0ec0e7af9e3baf0832.

- Tightened bounds on DtoType for JobAction
- Validate DTOs

sbliven committed:
Remove job-create interceptor

DTO validation is in the controller, so the interceptor isn't needed.

sbliven committed:
Add jsonpath-plus dependency

sbliven committed:
Add ValidateAction

Also tests and example

despadam committed:
Add fullfacet and fullquery for Jobs (#1422)

* add fullfacet and fullquery for jobs

* fix order of Get endpoints, fullfacet now returns empty array

* fix statusCode, statusMessage in tests

* add fullquery tests(wip) and endpoint examples, fix timestamps in job schema

* finish fullquery,fullfacet tests

* fix linting

sbliven committed:
Fix type error calling validate

sbliven committed:
Refactor job validate code

- Revert the previous commit, which was incorrect and incomplete
- Make valiateDTO and performActions generic so they can be shared
  between endpoints
- Separate out checkConfigVersion, which is specific to statusUpdate
- Move `getJobTypeConfiguration` calls up so this is only called once

sofyalaski committed:
fix issue 1411, empty jobParams allowed (#1433)

despadam committed:
Jobs: implement fullfacet (#1436)

* implement fullfacet

* tests

sbliven committed:
Add mocha tests for ValidateAction.

- Add new test for validate jobs to check for required parameters
  in both POST and PATCH operations
- Remove some unused TestData which didn't match the current Job DTOs
- Add 'validate' job to the test jobconfig.json

sbliven committed:
Add type checking to ValidateAction

- rename top-level array to 'request'. This leaves open the possibility
  to validate other things beside the DTO
- now configuration takes the format of `"JSONPath": JSON-Schema`. The
  schema part is validated with Avj. Note that JSON Schemas can be very
  concise for simple datatypes, so this doesn't require deep knowledge
  of how to write JSONSchema.
- Update unit & mocha tests

sbliven committed:
Lint fixes

- remove 'any' via typeguard
- `npm run lint:fix`
- remove duplicate (and outdated) docstring

sbliven committed:
Lint fixes

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Renumber tests within the 'Validate Job Action' section.

sbliven committed:
Merge branch 'release-jobs' into validateaction

(previous merge commit lost a parent)

sbliven committed:
Fix jobconfig.example.json to match ValidateAction

sbliven committed:
Merge remote-tracking branch 'main/master' into release-jobs

despadam committed:
Complete Jobs controller with datasetList checks (#1440)

* update controller

* update tests

* fix jobConfig test

* checkDatasetFiles should apply to all job types

sbliven committed:
Change example to use datasetList

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Whitespace changes

despadam committed:
Fix checkDatasetFiles after merging with master (#1459)

* fix checkDatasetFiles

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Fix test value (incorrect following merge)

sbliven committed:
Fix failing test

sbliven committed:
Merge pull request #1421 from sbliven/validateaction

feat(jobs): ValidateAction

sbliven committed:
dep: Update jsonpath-plus

This had a critical vulnerability

sbliven committed:
Merge pull request #1462 from sbliven/release-jobs

chore(deps): Update jsonpath-plus

sbliven committed:
dep: add js-yaml dependency

sbliven committed:
Parse jobconfig as yaml.

Example files are deliberately not converted in this commit to test
json parsing with js-yaml

sbliven committed:
Convert jobConfig files to YAML

JSON is still accepted as well, but documentation will be updated
to use YAML.

Move jobConfig.example.yaml to the top-level for better visibility.

sbliven committed:
fix lint errors

sbliven committed:
Merge pull request #1463 from sbliven/yamlconfig

feat(jobs): accept yaml configuration data

sbliven committed:
Merge branch 'master' into release-jobs

sbliven committed:
Merge remote-tracking branch 'main/master' into release-jobs

sbliven committed:
refactor: Implement JobConfig as a NestJS module (#1539)

* Complete refactor of job configuration as NestJS modules

- Developing new actions is documented in README.md for developers
- Move everything related to job config to src/config/job-config
- Major refactor of parsing from jobConfig.yaml to JobConfig objects
- Create module, interface, and creator for each JobAction
- Add handlebar-utils to help consistently applying templates to jobs
  and Dtos.
  - In test contexts the handlebars helpers may not be registered, so
    avoid using custom helpers or register them in the test
- Add options to LogJobAction. This is also good for templating tests
- Change MailService.sendMail to match nest's MailerService.sendMail
- Accept empty jobConfigurationFile as no jobs
- CaslModule now depends indirectly on the MailerModule. Thus all
  controllers need to either mock CaslAbilityFactory or add a mock
  MailerModule before importing CaslModule.
- Most tests disable jobConfig except for those that directly test it.
  These use test/config/jobconfig.yaml

sbliven committed:
Support dataset validation in ValidateAction

During a `create` operation, the `validate` action can take a
`datasets` option listing path/schema pairs that will be applied
to any datasets in jobParams.datasetList.

Datasets are fetched from the database during the DTO validation step
if needed.

Validation of archive/retrieve/public lifecycle properties is no longer
done automatically, but must be configured in the jobConfig file.

Details:
- Separate ValidateCreateJobAction (create) and ValidateJobAction (update)
- Fix 'Internal Server Error' from job controller if a session timed out.
  If the session times out then only public jobs will be visible.
- Update documentation for validate action
- Remove checkDatasetState from the job controller.
  This must now be configured with a validate action
- Add unit test
- Remove hard-coded job types and dataset states

despadam committed:
Jobs RabbitMQ Action connection info should come from ConfigService (#1575)

* extract rabbitmq connection info from configService

* update .env and jobConfig examples

* rewrite as a common module

sbliven committed:
Fix circular dependency

sofyalaski committed:
fix: delete_job_groups instance authorization (#1534)

## Description
When in .env delete-job-group was set to admin, flowing the
authorization logic, admin should be able to delete a job. Instead, this
user would get denied access.

## Motivation
fix this

## Fixes
user can delete a job based on delete-job-group

## Changes:
* In casl, the endpoint authorization lines for JobDelete are separated
into their own if/else clauses
* added admin to delete_job_group in tests workflow
* 1920 test shour return a successful status code
* test for getting are adapted to minus one deleted test. 

## Tests included

- [x] Included for each change/fix?
- [x] Passing? 

## Documentation
- [ ] swagger documentation updated (required for API changes)
- [ ] official documentation updated

sofyalaski committed:
Minor fix and better naming for variables (#1612)

<!--
Follow semantic-release guidelines for the PR title, which is used in
the changelog.

Title should follow the format `<type>(<scope>): <subject>`, where
- Type is one of:
build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test|BREAKING
CHANGE
- Scope (optional) describes the place of the change (eg a particular
milestone) and is usually omitted
- subject should be a non-capitalized one-line description in present
imperative tense and not ending with a period

See
https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines
for more details.
-->

## Description
Minor changes in variable names to improve documentation 

## Motivation
While creating documentation some variables created confusion.

## Fixes
<!-- Please provide a list of the issues fixed by this PR -->

* Bug fixed (#X)

## Changes:
* tests 1750-1860 response for successful calls changed to
SuccessfulGetStatusCode instead of Patch
* lines 20-93 renamed variables to better represent their meaning. Based
on job type and ownership info.
* newDataset renamed to newJob in each test

## Tests included

- [ ] Included for each change/fix?
- [ ] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [ ] swagger documentation updated (required for API changes)
- [ ] official documentation updated

### official documentation info
<!-- If you have updated the official documentation, please provide PR #
and URL of the updated pages -->

sbliven committed:
Use HTML for emailaction body

sofyalaski committed:
Refactor tests based on the "create" field in job config file (#1624)

<!--
Follow semantic-release guidelines for the PR title, which is used in
the changelog.

Title should follow the format `<type>(<scope>): <subject>`, where
- Type is one of:
build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test|BREAKING
CHANGE
- Scope (optional) describes the place of the change (eg a particular
milestone) and is usually omitted
- subject should be a non-capitalized one-line description in present
imperative tense and not ending with a period

See
https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines
for more details.
-->

## Description
`Jobs.js` was extremely long, this PR separates it into smaller files

## Tests included

- [ ] Included for each change/fix?
- [ ] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [ ] swagger documentation updated (required for API changes)
- [x] official documentation updated

### official documentation info
<!-- If you have updated the official documentation, please provide PR #
and URL of the updated pages -->

despadam committed:
RabbitMQ service: fix logging and error handling (#1581)

* fix logging and error handling

* optional rabbitMQservice initialization

sbliven committed:
Resolve DatasetsService at runtime.

DatasetsService is request scoped, so it can't be added on module init.
This also avoids the circular dependency from import datasetService
during module initialization.

sbliven committed:
Fix bug with dataset validation

Mongoose documents do some Proxy shenanigans that confuses JSONPath.
They should first be converted to simple objects.

sbliven committed:
Fix mocha tests for jobs following the validateaction changes

- Update error messages for archivable, retrievable, and public jobs.
  These used to be bespoke errors, but now follow the standard 'Invalid
  request' error message.
- Change dataset property validation to throw 409 errors, to be
  consistent with the previous behavior
- Add datasetList to many job DTOs that now validate it
- Set archivable/retrievable properly in test datasets

despadam committed:
Add tests for job actions (#1644)

* url and rabbitmq tests

* email unit test

* fix failing tests

* remove unused vars

* fix rabbitmq testing

* fix lint

* fix and test using default sender value for emails

sbliven committed:
Merge branch 'master' into release-jobs

sbliven committed:
Fix lint issues

sofyalaski committed:
wip: fix after merge, commit c729599

sofyalaski committed:
fix after merge: add accessGroup in jobs controller

sofyalaski committed:
fix configuration

sofyalaski committed:
wip:broken dependency

sofyalaski committed:
fixed after merge

sofyalaski committed:
merge again to add latest changes

sofyalaski committed:
restore direct dependency on ConfigModule

sofyalaski committed:
revert unnecessary casting and fix linting

sofyalaski committed:
fix unit tests by providing a casl mock

sbliven committed:
Add comments to validateaction

Also removes the type from ValidateFunction. The schema is applied to an
arbitrary element of the JobDto or DatasetClass, so the type returned
should be `unknown` (the default).

sbliven committed:
Merge pull request #2 from sofyalaski/release-jobs-fix-merge

Resolve additional conflicts from merging master into release-jobs & update again

sbliven committed:
chore: merge main into release-jobs (#1693)

<!--
Follow semantic-release guidelines for the PR title, which is used in
the changelog.

Title should follow the format `<type>(<scope>): <subject>`, where
- Type is one of:
build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test|BREAKING
CHANGE
- Scope (optional) describes the place of the change (eg a particular
milestone) and is usually omitted
- subject should be a non-capitalized one-line description in present
imperative tense and not ending with a period

See
https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines
for more details.
-->

## Description
<!-- Short description of the pull request -->
Merge main into release-jobs.

I am making a PR for review since there were a large number of
non-trivial conflicts to resolve

sbliven committed:
Merge branch 'release-jobs' into validateaction

sbliven committed:
Lint fix

sbliven committed:
feat: Validate dataset properties in jobs (#1473)

## Description
Allows jobs to validate properties of the dataset.

## Motivation
The `validate` action currently allows operators to enforce attributes
of the DTO.

However, many jobs depend on `jobParams.datasetLists` to link a job to
certain datasets. This allows validation of properties on datasets
associated with the dataset.

The motivating example would be checking datasetLifecycle properties for
certain jobs. These are currently hard-coded for special job types. This
would allow them to be applied to custom job types as well.

```
configVersion: v1.0
jobs:
  - jobType: custom_archive
    create:
      auth: "#datasetAccess"
      actions:
        - actionType: validate
          datasets:
            datasetlifecycle.archivable:
              const: true
```

## Changes:

* Add `datasets` property to validate actions (only in `create`
operations) to validate linked dataset properties

## Tests included

(WIP pending tests and documentation)

- [x] Included for each change/fix?
- [x] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

### official documentation info
Docs are included in
https://github.com/SciCatProject/documentation/pull/56

sofyalaski committed:
remove datasetsValidation fromJobClass (#1728)

* removes the datasetsValidation from jobInstance

* fixed linting

* remove comment

---------

Co-authored-by: Despina Adamopoulou <despoina.adamopoulou@empa.ch>

despadam committed:
Jobs: rename `statusUpdate` to `update` (#1460)

* rename statusUpdate to update

* linting

* fixes after merge

* fix config error

sbliven committed:
Merge branch 'master' into release-jobs

emigun commented on src/casl/casl-ability.factory.ts:
I think this file is kind of growing out of control. It would be nice to refactor it so that at least each endpoint has its own file.

Junjiequan replied:
I agree. I've found the code hard to read, and I've also heard complaints from others. 
I will find a time to refactor this.

emigun replied:
Lovely!

Junjiequan commented on src/config/job-config/actions/corejobactioncreators.module.ts:
```diff
@@ -0,0 +1,93 @@
+ * CREATE_JOB_ACTION_CREATORS and UPDATE_JOB_ACTION_CREATORS can be extended (eg by a
+ * plugin) with additional creators.
+ */
+@Module({
+  imports: [
+    EmailJobActionModule,
+    LogJobActionModule,
+    ValidateJobActionModule,
+    URLJobActionModule,
+    RabbitMQJobActionModule,
+  ],
+  providers: [
```
~~I can see that each module in the array will be initialized two times. One for `CREAT_JOB_ACTION` and the other for `UPDATE_JOB_ACTION`.~~  <-- Correcting my initial comment - NestJS modules are singletons and their providers are only initialized once.
If I want to use only RabbitMQJobAction and do not need the others, how can I disable them from initializing to reduce unnecessary initialization overhead

nitrosx replied:
Valid point!!!

sbliven replied:
You could disable them by providing different arrays to nestjs for the `CREATE_JOB_ACTION_CREATORS` and `UPDATE_JOB_ACTION_CREATORS` tokens. Currently `app.module.ts` imports `CoreJobActionCreators` statically. However, if we implement a plugin mechanism in the future (#862) then I would expect RabbitMQJobAction to be moved from core to a plugin, and `AppModule` would use some kind of dynamic import mechanism. It was decided this was too complicated for now, so all actions are included in the core.

Junjiequan replied:
If it takes much more time to implement then it's better to skip it for now and add it in the future.
NestJS recently introduced new concept of [Conditional module configuration](https://docs.nestjs.com/techniques/configuration#conditional-module-configuration) which might be useful for future reference

Junjiequan commented on src/common/rabbitmq/rabbitmq.service.ts:
```diff
@@ -0,0 +1,120 @@
+  private connection: Connection;
+  private channel: Channel;
+
+  constructor(private readonly configService: ConfigService) {
+    Logger.log("Initializing RabbitMQService.", "RabbitMQService");
+
+    const hostname = this.configService.get<string>("rabbitMq.hostname");
+    const port = this.configService.get<number>("rabbitMq.port");
+    const username = this.configService.get<string>("rabbitMq.username");
+    const password = this.configService.get<string>("rabbitMq.password");
+
+    if (!hostname || !port || !username || !password) {
+      Logger.error(
+        "RabbitMQ is enabled but missing one or more config variables.",
+        "RabbitMQService",
+      );
```
`onApplicationBootstrap()` under the [app.module.ts ](https://github.com/SciCatProject/scicat-backend-next/blob/master/src/app.module.ts) tries to connect to the rabbitMQ using the same variables, so there will be two instances of rabbitMQ connections once enabled.
That's some legacy code from the ESS that is no longer used but not removed. Would you prefer me to create new PR to remove that or just delete them from this PR?

nitrosx replied:
@Junjiequan please create a PR to this PR

Junjiequan commented on src/common/rabbitmq/rabbitmq.module.ts:
```diff
@@ -0,0 +1,10 @@
+import { Module } from "@nestjs/common";
+import { ConfigModule } from "@nestjs/config";
+import { RabbitMQService } from "./rabbitmq.service";
+
+@Module({
+  imports: [ConfigModule],
+  providers: [RabbitMQService],
+  exports: [RabbitMQService],
+})
+export class RabbitMQModule {}
```
RabbitMQModule is created but is not imported anywhere. Is it intentional?

nitrosx replied:
Same question as @Junjiequan

sbliven replied:
It's imported by `CoreJobActionCreators`

Junjiequan replied:
Searching through the source code it seems that [CoreJobActionCreators](https://github.com/SciCatProject/scicat-backend-next/blob/954516bc707eb2f941de22118e921b3f3487b93a/src/config/job-config/actions/corejobactioncreators.module.ts) imports `RabbitMQJobActionModule`
and [RabbitMQJobActionModule](https://github.com/SciCatProject/scicat-backend-next/blob/954516bc707eb2f941de22118e921b3f3487b93a/src/config/job-config/actions/rabbitmqaction/rabbitmqaction.module.ts) imports `CommonModule`. 
My guess is, create job action for RabbitMQ works since commonModule imports RabbitMQ service as a provider.

Junjiequan commented on src/common/common.module.ts:
```diff
@@ -1,8 +1,24 @@
+      },
+      inject: [ConfigService],
+    },
+  ],
```
Wouldn't it be better to import `RabbitMQmodule` here instead of setting `RabbitMQService` as a provider? Considering rabbitmq module is created but not used anywhere

sbliven replied:
CommonModule is a bit of a weird module. It is needed for services which are configured globally in app.module but which need to be imported by other modules.

That said, I don't really get why RabbitMQService needs to be created here rather than in its own module. @despadam do you remember the reason?

emigun commented on src/config/job-config/README.md:
```diff
@@ -0,0 +1,51 @@
+
+Implementing an Action requires four (short) files:
+
+1. `action.ts` contains a class implementing `JobAction`. It's constructor can take any
```
```suggestion
1. `action.ts` contains a class implementing `JobAction`. Its constructor can take any
```

emigun replied:
Or perhaps `The constructor...`

emigun commented on src/config/job-config/actions/urlaction/urlaction.ts:
```diff
@@ -0,0 +1,83 @@
+      );
+    }
+
+    // TODO do something with the response?
```
Forgotten?

emigun commented on src/jobs/dto/create-job.dto.ts:
```diff
@@ -1,30 +1,52 @@
+    type: Object,
+    required: true,
+    description: "Job's parameters as defined by job template in configuration",
+  })
```
Did you consider using the openapi cli plugin instead for ApiProperty? I think it is more maintainable.

nitrosx replied:
I agree with @emigun , although it can be done in a new PR

sbliven replied:
I agree, this should be done to bring release-jobs into alignment with @emigun's changes in main.

emigun commented on src/jobs/dto/update-job.dto.ts:
```diff
@@ -1,18 +1,30 @@
+  @ApiProperty({
+    type: String,
+    required: true,
+    description: "Updated job status code for the current status.",
```
Would be nice with the automatic swagger from openapi cli plugin here too.

emigun commented on src/jobs/schemas/job.schema.ts:
```diff
@@ -2,100 +2,134 @@ import { Prop, Schema, SchemaFactory } from "@nestjs/mongoose";
+    default: () => uuidv4(),
     description: "Globally unique identifier of a job.",
-    readOnly: true,
   })
```
Same for this file (openapi cli plugin)

emigun commented on test/DatasetFilter.js:
```diff
@@ -120,7 +120,7 @@ describe("0400: DatasetFilter: Test retrieving datasets using filtering capabili
       username: "archiveManager",
       password: TestData.Accounts["archiveManager"]["password"],
-    });
+    });  
```
Just a whitespace here

sbliven replied:
We should add js files to eslint. That will lead to a huge diff though, so let's do it after merging

emigun commented on test/DatasetLifecycle.js:
```diff
@@ -29,7 +29,7 @@ describe("0500: DatasetLifecycle: Test facet and filter queries", () => {
   });
 
-
+  
```
And whitespace here. At least that's what I think it means :)

emigun commented on test/DerivedDataset.js:
```diff
@@ -21,6 +21,7 @@ describe("0700: DerivedDataset: Derived Datasets", () => {
     accessTokenAdminIngestor = await utils.getToken(appUrl, {
       username: "adminIngestor",
       password: TestData.Accounts["adminIngestor"]["password"],
+
```
Remove this line?

emigun commented:
Wow! :star_struck: Honestly, this looks amazing! :star: 

Well done!! :1st_place_medal: :1st_place_medal: :1st_place_medal:

nitrosx commented on src/common/utils.ts:
```diff
@@ -819,6 +819,24 @@ export const addApiVersionField = <T extends object>(
+    jobResultObject: jobResultObject,
+  };
+};
+
```
I wonder if this is the right place for this function given that it is specific for the job subsystem?
Not a blocking issue

sbliven replied:
IMO it fits well with `addCreatedByFields` and `addUpdatedByField` in the same file, but I see your point.

nitrosx commented on src/users/users.service.ts:
```diff
@@ -230,6 +236,22 @@ export class UsersService implements OnModuleInit {
     return null;
   }
 
+  async findByUsername2JWTUser(username: string): Promise<JWTUser | null> {
+    const userIdentity = await this.userIdentityModel
+      .findOne({ "profile.username": username })
+      .exec();
+    if (userIdentity) {
+      const userProfile = userIdentity.profile;
+      return {
+        _id: userProfile.id,
+        username: userProfile.username,
+        email: userProfile.email,
+        currentGroups: userProfile.accessGroups,
+      } as JWTUser;
+    }
+    return null;
+  }
+
```
This function is already present in master.

sofyalaski replied:
Indeed, there's a very similar function that finds a user by ID.
I introduced this new function at a time because of the use case where a user in admin groups creates a job for another user. This other user is specified in the newly introduced optional ownerUser parameter in the job DTO.  It would expect an input as a username as of now. 
We can reuse the `findById2JWTUser` and then job DTO needs to provide an actual ID of the user for whom the job will be created. Should I create a PR for this? 

Now when I think about this use case, I think there is another check that is missing. If both ownerUser and ownerGroup are provided in DTO, we don’t check upon job creation, if the ownerUser is part of that ownerGroup. I will create a PR with that check. 

In most cases, ownerUser is optional and is set during job creation based on logged in user.

sofyalaski replied:
@nitrosx, Despina and I decided to keep the `findByUsername2JWTUser` function. As I explained, it is used to find a user provided in `ownerUser` in DTO. We thought that a userID could be used for that field, but concluded that it would be confusing since the created job would also contain just a userID. In a real case scenario, if the admin would want to look at the created job and see some userID in the ownerUser field, it will require a call to User backend to understand who this job was created for. 
And username is probably a unique identifier too.

nitrosx reviewed (commented):
Just few minor comments.
Good work.

sbliven committed:
Fix bug in validateaction

As reported by @console, adding a job that validated a dataset
property resulted in an error 'Invalid options for ValidateJobAction.'

Junjiequan committed:
refactor: remove RabbitMQ integration and related logic from App Module (#1754)

refactor(app): remove RabbitMQ integration and related logic from AppModule

despadam committed:
feat(jobs): fix ownerUser, contactEmail logic and datasetList validation (#1758)

* cherry-pick from commit 61299cd

* fix failing tests

* strengthen datasetList validation

* lint

* use enum in controller

* fix ownerUser

* fix ownerUser logic

* fix tests

sbliven committed:
fix: improve validate action validation (#1755)

## Description
Improve the error message when you misconfigure a 'validate' action in a
create job.

## Motivation
We ran into this error with a buggy jobConfig.yaml and had trouble
tracking it back. This improves the error message.

## Fixes
(no issue filed)

## Changes:

## Tests included

- [ ] Included for each change/fix?
- [x] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [ ] swagger documentation updated (required for API changes) - N/A
- [ ] official documentation updated

### official documentation info
<!-- If you have updated the official documentation, please provide PR #
and URL of the updated pages -->

sbliven commented:
TODO before merging
- [x] Resolve schema concerns from #1742 and merge it to this branch
- [x] Replace `@ApiProperty` annotations with the cli-plugin following #1590 and #1611 (PR: #1770)
- [x] Try [Conditional module configuration](https://docs.nestjs.com/techniques/configuration#conditional-module-configuration) for the RabbitMqModule (PR: #1771)
- [x] Minor lint and comment changes
- [x] Merge in all the bump PRs committed despite the code freeze

sbliven committed:
Small fixes responding to review comments

- Log the response if a url action returns an error
- Remove trailing whitespace from js files

sbliven committed:
Merge remote-tracking branch 'main/master' into release-jobs

despadam commented:
PRs:
- Replace `ApiProperty` annotations with the openapi cli-plugin: https://github.com/SciCatProject/scicat-backend-next/pull/1770
- Conditional RabbitMQModule: https://github.com/SciCatProject/scicat-backend-next/pull/1771

despadam committed:
feat(jobs): register rabbitMQ as a conditional module (#1771)

register rabbitMQ as a conditional module in app.module

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs

despadam committed:
feat(jobs): replace ApiProperty with openapi cli-plugin (#1770)

* replace ApiProperty with openapi cli-plugin

* make jobResultObject required

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs

sofyalaski committed:
fix: permission for ownerUser to read jobs (#1774)

* fix permission for ownerUser to read jobs;
extract email of admin in testing

* fix linting

* requested changes

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs

despadam committed:
feat(jobs): Ensure backwards compatibility (#1742)

* make jobs backwards compatible

* - re-introduce v3 DTOs
- update create/update interceptors and enforce types
- ensure contactEmail does not get overwritten
- add tests for api/v3/jobs

* fix linting

* fix linting

* fix failing tests

* fix v3 return types

* refactor interceptors, fix job update, map executionTime correctly

* fix linting

* optional emailJobInitiator, fix jobResultObject

* update dtos and tests

* fix tests

* strengthen datasetList validation

* update JobClassV3

* lint

* use enum in controller

* finish tests

* improve test descriptions

* fix anonymous tests

* fix testing

* finish tests

* fix utils and service for job update

* add tests for anonymous jobs from normal user

* fix job.v3 schema

* job by anonymous user

* rename JobClassV3 to OutputJobV3Dto and fix ownerGroup compatibility logic

* add descriptions in dtos

* fix tests after merging

* fix test number

despadam committed:
Merge branch 'master' into release-jobs

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs

despadam committed:
split v3,v4 for fullfacet,delete

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs

despadam committed:
fix details in 3 tests

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs

despadam committed:
revert change in failing test

despadam committed:
Merge branch 'SciCatProject:release-jobs' into release-jobs
